core/src/test/scala/org/apache/spark/sql/delta/rowid/RowIdSuite.scala

```diff
@@ -57,4 +63,236 @@ class RowIdSuite extends QueryTest
+          .select((col("id") + 10000L).as("row_id"))
+          .write.format("delta").save(dir.getAbsolutePath)
+
+        checkAnswer(
+          spark.read.load(dir.getAbsolutePath),
+          (0 until 1000).map(i => Row(i + 10000L))
+        )
+      }
+    }
+  }
+
+  test("Throws error when assigning row IDs without stats") {
+    withSQLConf(
+      defaultRowIdFeatureProperty -> "supported",
+      DeltaSQLConf.DELTA_COLLECT_STATS.key -> "false") {
+      withTable("target") {
+        val err = intercept[UnsupportedOperationException] {
+          spark.range(end = 10).write.format("delta").saveAsTable("target")
+        }
+        assert(err.getMessage === "Cannot assign row IDs without row count statistics.")
+      }
+    }
+  }
+
+  test(s"CONVERT TO DELTA assigns row ids") {
+    withRowIdsEnabled(enabled = true) {
+      withTempDir { dir =>
+        spark.range(10)
+          .write.format("parquet").mode("overwrite").save(dir.getAbsolutePath)
+
+        sql(s"CONVERT TO DELTA parquet.`$dir`")
+
+        val log = DeltaLog.forTable(spark, dir)
+        assertRowIdsAreValid(log)
+      }
+    }
+  }
+
+  test(s"CONVERT TO DELTA throws error when converting without statistics") {
+    withRowIdsEnabled(enabled = true) {
+      withTempDir { dir =>
+        spark.range(10)
+          .write.format("parquet").mode("overwrite").save(dir.getAbsolutePath)
+
+        val err = intercept[UnsupportedOperationException] {
+          sql(s"CONVERT TO DELTA parquet.`$dir` NO STATISTICS")
+        }
+        assert(err.getMessage === "Cannot assign row IDs without row count statistics.")
+      }
+    }
+  }
+
+  test("manually setting row ID high watermark is not allowed") {
+    withRowIdsEnabled(enabled = true) {
+      withTempDir { dir =>
+        spark.range(start = 0, end = 1000, step = 1, numPartitions = 10)
+          .write.format("delta").save(dir.getAbsolutePath)
+
+        val log = DeltaLog.forTable(spark, dir)
+
+        val exception = intercept[IllegalStateException] {
+          log.startTransaction().commit(Seq(
+            RowIdHighWaterMark(highWaterMark = 9001, preservedRowIds = false)),
+            ManualUpdate)
+        }
+        assert(exception.getMessage.contains(
+          "Manually setting the Row ID high water mark is not allowed"))
+      }
+    }
+  }
+
+  test("ALTER TABLE cannot enable Row IDs on existing table") {
+    withRowIdsEnabled(enabled = false) {
+      withTable("tbl") {
+        spark.range(10).write.format("delta").saveAsTable("tbl")
+
+        val log = DeltaLog.forTable(spark, TableIdentifier("tbl"))
+        assert(!RowId.rowIdsEnabled(log.update().protocol, log.update().metadata))
+
+        val err = intercept[UnsupportedOperationException] {
+          sql(s"ALTER TABLE tbl SET TBLPROPERTIES ('${DeltaConfigs.ROW_IDS_ENABLED.key}' = true)")
+        }
+        assert(err.getMessage === "Cannot enable Row IDs on an existing table.")
+      }
+    }
+  }
```

Comment:
is it also possible to test the conflict resolution? What about disabling when enabled before?

Reply:
Conflict is pretty hard to test, unfortunately. I did add a test to check that disabling row ids works though